opentelemetry-sdk/src/opentelemetry/sdk/_metrics/metric_reader_storage.py

```diff
@@ -12,23 +12,101 @@
 
-from typing import Iterable
+from threading import RLock
+from typing import Dict, Iterable, List
 
-from opentelemetry.sdk._metrics.aggregation import AggregationTemporality
+from opentelemetry._metrics.instrument import Instrument
+from opentelemetry.sdk._metrics.aggregation import AggregationTemporality, LastValueAggregation
 from opentelemetry.sdk._metrics.measurement import Measurement
 from opentelemetry.sdk._metrics.point import Metric
 from opentelemetry.sdk._metrics.sdk_configuration import SdkConfiguration
+from opentelemetry.sdk._metrics.view import View
+from opentelemetry.sdk._metrics._view_instrument_match import _ViewInstrumentMatch
 
 
-# TODO: #2378
 class MetricReaderStorage:
     """The SDK's storage for a given reader"""
 
     def __init__(self, sdk_config: SdkConfiguration) -> None:
-        pass
+        self._lock = RLock()
+        self._sdk_config = sdk_config
+        self._view_instrument_match: Dict[Instrument, List[_ViewInstrumentMatch]] = {}
+
+    def _get_or_init_view_instrument_match(
+        self, instrument: Instrument
+    ) -> List["_ViewInstrumentMatch"]:
+        # Optimistically get the relevant views for the given instrument. Once set for a given
+        # instrument, the mapping will never change
+        if instrument in self._view_instrument_match:
+            return self._view_instrument_match[instrument]
+
+        with self._lock:
+            # double check if it was set before we held the lock
+            if instrument in self._view_instrument_match:
+                return self._view_instrument_match[instrument]
+
+            # not present, hold the lock and add a new mapping
+            matches = []
+            for view in self._sdk_config.views:
+                if view.match(instrument):
+                    # Note: if a view matches multiple instruments, this will create a separate
+                    # _ViewInstrumentMatch per instrument. If the user's View configuration includes a
+                    # name, this will cause multiple conflicting output streams.
+                    matches.append(
+                        _ViewInstrumentMatch(
+                            name=view.name or instrument.name,
+                            resource=self._sdk_config.resource,
+                            instrumentation_info=None,
+                            aggregation=view.aggregation,
+                            unit=instrument.unit,
+                            description=view.description,
+                        )
+                    )
+
+            # if no view targeted the instrument, use the default
+            if not matches:
+                matches.append(
+                    _ViewInstrumentMatch(
+                        resource=self._sdk_config.resource,
+                            instrumentation_info=None,
+                            aggregation=LastValueAggregation, # TODO: this needs to be set to the aggregation on the instrument
```

Comment:
I'm not sure what the right thing to do is here. Passing in a class doesn't work for any classes that require additional initialization parameters. For example, the `SumAggregation` needs to know about monotonicity and instrument aggregation temporality.

It might be best to set an aggregation attribute on the instrument and use that instead?

Reply:
Yes, good point, I think I can refactor :v:

Reply:
> Passing in a class doesn't work for any classes that require additional initialization parameters.

+1 and some of the configuration parameters can also come from the View e.g. bucket boundaries for Histograms

> It might be best to set an aggregation attribute on the instrument and use that instead?

I don't think that would work since this field is None unless a View is given that specifies the `aggregation`. In the cases where it's None, you would do the default per instrument.

There is an open discussion for how to configure this in Diego's other PR https://github.com/open-telemetry/opentelemetry-python/pull/2391#discussion_r798932435. I think one good option would be something like this

```py
from opentelemetry.sdk._metrics.view import SumFactory, ExplicitBucketHistogramFactory, ...

views = [
  View(
    ...
    aggregation_factory=SumFactory(),
  ),
  View(
    ...,
    aggregation_factory=ExplicitBucketHistogramFactory(
      explicit_bounds=[1, 4, 16, 64],
   ),
  ),
]

meter_provider = MeterProvider(views=views, ...)
```

where the specific aggregation factories bridge to the internal aggregation classes
```
from opentelemetry.sdk._metrics.aggregation import Aggregation, ExplicitBucketHistogramAggregation

class _AggregationFactory(ABC):
  @abstractmethod
  def create_aggregation(instrument: Instrument) -> Aggregation:
    """Creates an aggregation for the given instrument"""

class HistogramFactory(_AggregationFactory):
  def __init__(self, explicit_bounds):
    self._explicit_bounds = explicit_bounds

  def create_aggregation(insturment: Instrument) -> Aggregation:
    return ExplicitBucketHistogramAggregation(self._explicit_bounds)

...

Any other ideas @codeboten ?
```

Reply:
@aabmass 
Does this mean we let the users decide on what the default behaviors would be?

Reply:
@lzchen I'm confused, how does this effect the defaults?

I believe there is a "MAY" or "SHOULD" in the spec that users could change the defaults. I think that would be separate from the View configuration though

Reply:
I updated the PR to use a default for each instrument type, PTAL.

Reply:
@aabmass 
Maybe I misunderstood what the aggregation factory classes are. In your code example, you are letting users define a factory when they instantiate a View right? Doesn't that mean you are given them control over what type of aggregation + configuration per aggregation is used for specific instruments?

Reply:
Right, that is part of the View spec, they can override the aggregation for a specific view. If they don't specify an aggregation though, it will still use the default for that instrument